### DISCUSSION TITLE
Learned how easy it is to make something subscriptable with generics

### DISCUSSION THREAD

**Discussion by Moosems** [beartype/beartype#464] [Show and tell]
@leycec Am I doing it wrong or is it really this non-trivial to subscript my types?!

```python3
from beartype.typing import TypeVar, Generic

_T = TypeVar("_T")

class Something(Generic[_T]):
    data: _T
    
    def __init__(self, data: _T = None) -> None:
        self.data = data
        
s = Something[int](5)
s.data # Type checker shows int! Even when I obscure my int value!
```

**Comment by leycec** [comment]
**lol.** Yup. It's even easier than that under Python ≥ 3.12, where you no longer need to explicitly declare your **type variables** (i.e., `TypeVar()` objects). Instead, just subscript your types by your type variables and away you go: e.g.,

```python
from beartype.typing import Generic

class Something[_T](Generic[_T]):  # <-- look, ma! no "TypeVar"
    data: _T
    
    def __init__(self, data: _T = None) -> None:
        self.data = data
        
s = Something[int](5)
s.data # Type checker shows int! Even when I obscure my int value!
```

The runtime type-checking issue, of course, is that *no* runtime type-checker (including @beartype) propagates generic subscriptions. That is, @beartype currently fails to "copy" the `int` type hint subscripting `Something[int]` into either the `data: int` field or the `def __init__(self, data: int = None) -> None:` constructor. The `int` is entirely ignored at runtime.

In fact, the core `typing.Generic` superclass goes out of its way to destroy the `int` type hint subscripting `Something[int]` at runtime. Any runtime type-checker (including @beartype) trying to type-check generic subscriptions thus needs to do CrazyStuff™ like monkey-patching the core `typing.Generic` superclass (e.g., to record generic subscriptions in some sort of instance variable like `__beartype_child_hints__`). Probably *not* a great idea... but what you gonna do, huh? :thinking: 

@beartype: We didn't make crazy. We just monkey-patch the crazy away.
